actors/builtin/market/market_actor.go

```diff
@@ -181,94 +180,160 @@ func (a Actor) PublishStorageDeals(rt Runtime, params *PublishStorageDealsParams
 	if !callerOk {
 		rt.Abortf(exitcode.ErrForbidden, "caller %v is not worker or control address of provider %v", caller, provider)
 	}
-
 	resolvedAddrs := make(map[addr.Address]addr.Address, len(params.Deals))
 	baselinePower := requestCurrentBaselinePower(rt)
 	networkRawPower, networkQAPower := requestCurrentNetworkPower(rt)
 
-	var newDealIds []abi.DealID
+	// Drop invalid deals
 	var st State
+	proposalCidLookup := make(map[cid.Cid]struct{})
+	validProposalCids := make([]cid.Cid, 0)
+	validDeals := make([]ClientDealProposal, 0, len(params.Deals))
+	totalClientLockup := make(map[addr.Address]abi.TokenAmount)
+	totalProviderLockup := abi.NewTokenAmount(0)
+
+	validInputBf := bitfield.New()
+	rt.StateReadonly(&st)
+	msm, err := st.mutator(adt.AsStore(rt)).withPendingProposals(ReadOnlyPermission).
+		withEscrowTable(ReadOnlyPermission).withLockedTable(ReadOnlyPermission).build()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load state")
+	for di, deal := range params.Deals {
+		/*
+			drop malformed deals
+		*/
+		if err := validateDeal(rt, deal, networkRawPower, networkQAPower, baselinePower); err != nil {
+			rt.Log(rtt.INFO, "invalid deal %d: %s", di, err)
+			continue
+		}
+		if deal.Proposal.Provider != provider && deal.Proposal.Provider != providerRaw {
+			rt.Log(rtt.INFO, "invalid deal %d: cannot publish deals from multiple providers in one batch", di)
+			continue
+		}
+		client, ok := rt.ResolveAddress(deal.Proposal.Client)
+		if !ok {
+			rt.Log(rtt.INFO, "invalid deal %d: failed to resolve proposal.Client address %v for deal ", di, deal.Proposal.Client)
+			continue
+		}
+
+		/*
+			drop deals with insufficient lock up to cover costs
+		*/
+		if _, ok := totalClientLockup[client]; !ok {
+			totalClientLockup[client] = abi.NewTokenAmount(0)
+		}
+		totalClientLockup[client] = big.Sum(totalClientLockup[client], deal.Proposal.ClientBalanceRequirement())
+		clientBalanceOk, err := msm.balanceCovered(client, totalClientLockup[client])
```

Comment:
I see that this is the same as the current checks (the deletions below), but I don't understand the address handling here. Based on an examination of the `AddBalance` method, it appears that the balanceTables only get populated with ID addresses.

But we aren't resolving addresses here (or in the `balanceCovered` method), so this is...gonna fail if the proposal has robust addresses for the client / provider?

I'm probably missing something.

Reply:
EDIT: I just noticed that the `balanceCovered` method is new -- in that case i think there's a bug here

Reply:
hmm does the client address resolution on line 212 address your concerns?

Reply:
oh, yeah, that does it...that line was probably invisible last night